builds.go

```diff
@@ -0,0 +1,293 @@
+					if from == "scratch" {
+						continue
+					}
+					var resolved *RemoteResolvedFull
+					if parent.SourceID != nil {
+						sourceArchResolvedMutex.RLock()
+						resolvedFunc, ok := sourceArchResolved[*parent.SourceID+"-"+build.Build.Arch]
+						if !ok {
+							panic("parent of " + source.SourceID + " on " + build.Build.Arch + " should be " + *parent.SourceID + " but that sourceId is unknown to us!")
+						}
+						sourceArchResolvedMutex.RUnlock()
+						resolved = resolvedFunc()
+					} else {
+						lookup := from
+						if parent.Pin != nil {
+							lookup += "@" + *parent.Pin
+						}
+
+						resolved, err = resolveRemoteArch(context.TODO(), lookup, build.Build.Arch)
+						if err != nil {
+							panic(err)
+						}
+					}
+					if resolved == nil {
+						fmt.Fprintf(os.Stderr, "%s (%s) -> not yet! [%s]\n", source.SourceID, source.AllTags[0], build.Build.Arch)
+						close(outChan)
+						return nil
+					}
+					build.Build.ResolvedParents.Set(from, *resolved)
+					build.Build.Parents.Set(from, string(resolved.Manifest.Desc.Digest))
+				}
+
+				// buildId calculation
+				buildIDJSON, err := json.Marshal(&build.Build.BuildIDParts)
+				if err != nil {
+					panic(err)
+				}
+				buildIDJSON = append(buildIDJSON, byte('\n')) // previous calculation of buildId included a newline in the JSON, so this preserves compatibility
+				// TODO if we ever have a bigger "buildId break" event (like adding major base images that force the whole tree to rebuild), we should probably ditch this newline
+
+				build.BuildID = fmt.Sprintf("%x", sha256.Sum256(buildIDJSON))
+				fmt.Fprintf(os.Stderr, "%s (%s) -> %s [%s]\n", source.SourceID, source.AllTags[0], build.BuildID, build.Build.Arch)
+
+				build.Build.Img = strings.ReplaceAll(strings.ReplaceAll(stagingTemplate, "BUILD", build.BuildID), "ARCH", build.Build.Arch) // "oisupport/staging-amd64:xxxx"
+
+				build.Build.Resolved, err = resolveRemoteArch(context.TODO(), build.Build.Img, build.Build.Arch)
+				if err != nil {
+					panic(err)
+				}
+
+				json, err := json.Marshal(&build)
+				if err != nil {
+					panic(err)
+				}
+				outChan <- out{
+					buildId: build.BuildID,
+					json:    json,
+				}
+
+				return build.Build.Resolved
+			})
+			sourceArchResolvedMutex.Lock()
+			sourceArchResolved[source.SourceID+"-"+build.Build.Arch] = sourceArchResolvedFunc
+			sourceArchResolvedMutex.Unlock()
+			go sourceArchResolvedFunc()
+		}
+
+		if err := stdout.Close(); err != nil {
+			panic(err)
+		}
+		if err := jq.Wait(); err != nil {
+			panic(err)
+		}
+
+		close(outs)
+	}()
+
+	fmt.Print("{")
+	first := true
+	for outChan := range outs {
+		out, ok := <-outChan
+		if !ok {
+			continue
+		}
+		if !first {
+			fmt.Print(",")
+		} else {
+			first = false
+		}
+		fmt.Println()
+		buildId, err := json.Marshal(out.buildId)
+		if err != nil {
+			panic(err)
+		}
+		fmt.Printf("\t%s: %s", string(buildId), string(out.json))
+	}
+	fmt.Println()
```

Comment:
Why newline here? You do not appear to be be pretty printing otherwise.

Reply:
I'm going to answer this one and "Why the tab character?" together -- I _am_ roughly pretty-printing!  The value objects here are really large and deeply nested, so they are a _lot_ when they're pretty-printed to the terminal, but it _is_ useful to have the entries of the outer map be line-based, so this code has output that looks roughly like this:

```json
{
	"buildId-here": { ... large object here, on a single line ... },
	...
}
```

In other words, the load-bearing output thus resembles pretty closely the progress output (one line per build).

We pretty print the end result outside this for better diffs in Git (because Git's diffing algorithms generally do a better job with more newlines vs less).

Reply:
I forgot `out.json` had a newline at the end, so I thought everything ended up on a single line, with tabs between the key and the previous value. But now that I remember the JSON has a newline at the end, the `fmt.Println()` is _not_ necessary, right?

So the output ends up looking like this?

```json
{	"build-id-0": {…}
,	"build-id-1": {…}
…
,	"build-id-n": {…}

}
```

Reply:
Nope - the `fmt.Printf` above does _not_ include a newline explicitly so that `,` can be on the line above (because that looks much cleaner / ergonomic):

```console
$ BASHBREW_STAGING_TEMPLATE=foo:BUILD ./builds <(printf '{}')
{
}
$ BASHBREW_STAGING_TEMPLATE=tianon/zz-staging:ARCH-BUILD ./builds <(BASHBREW_ARCH_NAMESPACES= BASHBREW_NAMESPACE=tianon BASHBREW_LIBRARY="$HOME/docker/tianon-bashbrew/library" ./sources.sh true 2>/dev/null) 2>/dev/null
{
	"93918980eb5f67dc11efc857eba1f0a33d9f433485d3d6b16101fb81cff85bb7": {"buildId":"93918980eb5f67dc11efc857eba1f0a33d9f433485d3d6b16101fb81cff85bb7","build":{"img":"tianon/zz-staging:amd64-93918980eb5f67dc11efc857eba1f0a33d9f433485d3d6b16101fb81cff85bb7","resolved":{"manifest":{"ref":"tianon/zz-staging:amd64-93918980eb5f67dc11efc857eba1f0a33d9f433485d3d6b16101fb81cff85bb7@sha256:0d18be811527e931c717c17bd130b37cd63cc15861502a1ee623cf6108aee783","desc":{"mediaType":"application/vnd.docker.distribution.manifest.v2+json","digest":"sha256:0d18be811527e931c717c17bd130b37cd63cc15861502a1ee623cf6108aee783","size":524,"platform":{"architecture":"amd64","os":"linux"}}}},"sourceId":"f5f5c9e6e11fd02cff53245b41366e4e4b1c5a1c043a1d96ff4eccdb19194abc","arch":"amd64","parents":{},"resolvedParents":{}},"source":{"sourceId":"f5f5c9e6e11fd02cff53245b41366e4e4b1c5a1c043a1d96ff4eccdb19194abc","reproducibleGitChecksum":"dd443a9eeed44b2f73674af0954ae79ee4efd1b65dfbe35f5911914ec8b1bf18","allTags":["tianon/true:latest"],"entry":{"GitRepo":"https://github.com/tianon/dockerfiles.git","GitFetch":"refs/heads/master","GitCommit":"bfa71889cd8e48e7a1d2632422785c479bd9bd45","Directory":"true","File":"Dockerfile","Builder":"","SOURCE_DATE_EPOCH":1675234271},"arches":{"amd64":{"tags":["tianon/true:latest"],"archTags":[],"froms":["scratch"],"platformString":"linux/amd64","platform":{"architecture":"amd64","os":"linux"},"parents":{"scratch":{"sourceId":null,"pin":null}}}}}},
	"ea8772845de98e529b90034e19eab1979ebbe697963cbafa97c823d06eb32fa8": {"buildId":"ea8772845de98e529b90034e19eab1979ebbe697963cbafa97c823d06eb32fa8","build":{"img":"tianon/zz-staging:amd64-ea8772845de98e529b90034e19eab1979ebbe697963cbafa97c823d06eb32fa8","resolved":{"manifest":{"ref":"tianon/zz-staging:amd64-ea8772845de98e529b90034e19eab1979ebbe697963cbafa97c823d06eb32fa8@sha256:125a236c3f02370d5110e7dce47664d99f4f4a3da00151f0bf9fb8568be4e398","desc":{"mediaType":"application/vnd.docker.distribution.manifest.v2+json","digest":"sha256:125a236c3f02370d5110e7dce47664d99f4f4a3da00151f0bf9fb8568be4e398","size":524,"platform":{"architecture":"amd64","os":"linux"}}}},"sourceId":"ffc37b0864706846f1f9f014df57ac0572ab9dd9eb99f5f0d318b26cd19f4721","arch":"amd64","parents":{},"resolvedParents":{}},"source":{"sourceId":"ffc37b0864706846f1f9f014df57ac0572ab9dd9eb99f5f0d318b26cd19f4721","reproducibleGitChecksum":"203c5d023e44b5df1f46d41df253a770ed508c672738625b91ff3f464193fbdb","allTags":["tianon/true:yolo"],"entry":{"GitRepo":"https://github.com/tianon/dockerfiles.git","GitFetch":"refs/heads/master","GitCommit":"47fe0d0cb5ede5a36907162c25b058fd926f22fc","Directory":"true","File":"Dockerfile.yolo","Builder":"","SOURCE_DATE_EPOCH":1677043314},"arches":{"amd64":{"tags":["tianon/true:yolo"],"archTags":[],"froms":["scratch"],"platformString":"linux/amd64","platform":{"architecture":"amd64","os":"linux"},"parents":{"scratch":{"sourceId":null,"pin":null}}}}}}
}
```

Reply:
My internal compiler toolchain is a bit off. I missed https://github.com/docker-library/meta-scripts/pull/5/files#diff-293e1f9ac7fe0e3a8531eedbc7cd2b8772bc0dda2ae9497f288bfa1a6cc4b3f4R284 , and it looks like https://github.com/docker-library/meta-scripts/pull/5/files#diff-293e1f9ac7fe0e3a8531eedbc7cd2b8772bc0dda2ae9497f288bfa1a6cc4b3f4R232 does not end up the this JSON.